### ISSUE TITLE
ansible_winrm_server_cert_validation=ignore doesn't work when use delegate_to run role in a wodows server

### ISSUE THREAD

**Issue by lsbwant** [ansible/ansible#32673]
ISSUE TYPE:
 - Bug Report

COMPONENT NAME:
delegate_to

ANSIBLE VERSION:
2.1 2.2 2.3 2.4

CONFIGURATION
```
DEFAULT_TIMEOUT(/etc/ansible/ansible.cfg) = 60
HOST_KEY_CHECKING(/etc/ansible/ansible.cfg) = False
```

OS / ENVIRONMENT:
Redhat7.3 in AWS
python 2.7.5
windows server 2012R2 in AWS

SUMMARY:
ansible_winrm_server_cert_validation=ignore doesn't work when I configure it in an inventory file then use directive "delegate_to" run a role in windows servers. the error message like following:

```
TASK [schema_drop_sqlserver : Database job executing.. please with..] *****************************************************************************************************************************
task path: /home/ec2-user/ShuaibingLi_datapump1.3_sql_2014/roles/schema_drop_sqlserver/tasks/main.yml:5
<52.220.55.74> ESTABLISH WINRM CONNECTION FOR USER: Administrator on PORT 5986 TO 52.220.55.74
fatal: [localhost]: UNREACHABLE! => {
    "changed": false,
    "msg": "ssl: HTTPSConnectionPool(host='52.220.55.74', port=5986): Max retries exceeded with url: /wsman (Caused by SSLError(SSLError(1, u'[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:579)'),))",
    "unreachable": true
}
```

It will work if I don't use directive "delegate_to", just show a warning in console like following:

```
TASK [schema_drop_sqlserver : Database job executing.. please with..] **************************************************************************
task path: /home/jenkins/colline-ansible/roles/schema_drop_sqlserver/tasks/main.yml:5
<52.220.55.74> ESTABLISH WINRM CONNECTION FOR USER: Administrator on PORT 5986 TO 52.220.55.74
/usr/lib/python2.7/site-packages/urllib3/connectionpool.py:858: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  InsecureRequestWarning)
EXEC (via pipeline wrapper)
```

Or definite ansible_winrm_server_cert_validation=ignore in playbook rather than inventory file, it is work too

STEPS TO REPRODUCE:

hosts_inventory file:
```
[MSDatabaseServers]
192.168.1.100 ansible_user=Administrator ansible_password=password@111 ansible_port=5986 ansible_connection=winrm ansible_winrm_server_cert_validation=ignore
```

database_server.yml
```
---
db_server_ip: "192.168.1.100"
db_schema_name: "test_110"
```

importDatabase.yml
```
---
- name: Import Database Schema
  vars_files:
    - database_server.yml
  hosts: localhost
  connection: local
  any_errors_fatal: true
  serial: 1

  roles:
    - { role: db_import_sqlserver, delegate_to: "{{db_server_ip}}", when: "db_server_type == 'sqlserver'" }
```

roles/db_import_sqlserver/tasks/main.yml
```
---
  - name: STARTING DATABASE IMPORT FOR SQLSERVER
    debug: msg="ON DATABASE {{ db_server_ip }} with DB SCHEMA {{db_schema_name}}..."

  - name: "Database job executing.. please with.."
    raw: dir
```

ansible-playbook -i hosts_inventory importDatabase.yml

EXPECTED RESULTS:
have no CERTIFICATE_VERIFY_FAILED error

ACTUAL RESULTS:
```
fatal: [localhost]: UNREACHABLE! => {
    "changed": false,
    "msg": "ssl: HTTPSConnectionPool(host='192.168.1.100', port=5986): Max retries exceeded with url: /wsman (Caused by SSLError(SSLError(1, u'[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:579)'),))",
    "unreachable": true
}
```

If add ansible_winrm_server_cert_validation: ignore in importDatabase, the yml file will run successful.

importDatabase.yml
```
---
- name: Import Database Schema
  vars_files:
    - database_server.yml
 vars:
    - ansible_winrm_server_cert_validation: ignore

  hosts: localhost
  connection: local
  any_errors_fatal: true
  serial: 1

  roles:
    - { role: db_import_sqlserver, delegate_to: "{{db_server_ip}}", when: "db_server_type == 'sqlserver'" }
```

**Comment by lsbwant** [comment]
Anyone like have a look?

**Comment by jborean93** [comment]
When Ansible has a `delegate_to`, what happens is it replaces the connection vars with the delegated host's connection vars. In the case of WinRM, pretty much all the `ansible_winrm_*` vars are not seen as connection vars and just as host facts. So when it comes to delegating to the Windows hosts, things like `ansible_winrm_server_cert_validation`, `ansible_winrm_transport` are not set as delegate_to uses the host facts/vars of the non delegated hosts.

There is work in the pipeline to fix this up and set all `ansible_winrm_*` vars as connection plugins but unfortunately it is still a WIP so isn't ready yet. In the meantime you can manually set the variables as a task var like so

```
- win_command: hostname
  delegate_to: windows-server
  vars:
    ansible_winrm_server_cert_validation: ignore
```

This way, Ansible will add the `ansible_winrm_server_cert_validation` to that task executation. While more verbose you can also do this as a workaround and source the var from your Windows host setting

```
- win_command: hostname
  delegate_to: windows-server
  vars:
    ansible_winrm_server_cert_validation: "{{hostvars['windows-server']['ansible_winrm_server_cert_validation']|default('ignore')}}"
```

This does the same as the first example but sources the value from the value set on your Windows host and use a default of `ignore` if it isn't set.

**Comment by lsbwant** [comment]
Thanks @jborean93 
I am strange it is worked in ubuntu 14.04 LTS. I launched Tower server from official AWS image last year. it is ubuntu 14.04 LTS, ansible version is 2.0. then I upgrade ansible to 2.1.  the image id is "ansible-tower-568ea79f-6747-f9eb-c698-c152b9819ec1 (ami-331cd150)".

OS/ENVIRONMENT:
ubuntu@ip-10-81-2-116:~$ ansible --version
ansible 2.1.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
ubuntu@ip-10-81-2-116:~$ python --version
Python 2.7.6
![capture](https://user-images.githubusercontent.com/3376664/32585745-e8aebc3c-c539-11e7-9176-f7f9cbf747bc.PNG)

**Comment by bgstack15** [comment]
I found a workaround, which fits with @jborean93's information. You need a task to set the fact you want, on the host in question! It's just a `set_fact`.

    - set_fact:
        ansible_winrm_server_cert_validation: ignore
    
    - win_ping:
      delegate_to: "{{ win_host }}"
      vars:
        ansible_user: "{{ win_user }}"
        ansible_port: 5986

And then your task should continue, albeit with that InsecureRequestWarning.

    TASK [certreq : win_ping] ******************************************************************************************
    task path: /etc/ansible/roles/certreq/tasks/1_certreq.yml:6
    Using module file /usr/lib/python2.7/site-packages/ansible/modules/windows/win_ping.ps1
    <win_host> ESTABLISH WINRM CONNECTION FOR USER: makecert on PORT 5986 TO win_host
    /usr/lib/python2.7/site-packages/urllib3/connectionpool.py:858: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
      InsecureRequestWarning)
    EXEC (via pipeline wrapper)
    ok: [linux_host -> win_host] => {
        "changed": false,
        "ping": "pong"
    }
    META: ran handlers
    META: ran handlers

**Comment by jborean93** [comment]
This issue has been fixed with https://github.com/ansible/ansible/commit/23b1dbacaf8a20f7147de0f67f8d6c67de2142ef and will be released with 2.5. You no longer need to explicitly set `ansible_winrm_server_cert_validation` on the task or to a non Windows host for this to work. This fix also works with any other winrm connection variables that are defined here https://docs.ansible.com/ansible/devel/plugins/connection/winrm.html

**Comment by lsbwant** [comment]
@jborean93 It is let me so excited. That's so great! Thank you very much!
